reactive/zookeeper.py

```diff
@@ -0,0 +1,60 @@
+'''
+Reactive handlers for Zookeeper.
+'''
+
+from charmhelpers.core import hookenv
+from charms.layer.zookeeper import Zookeeper
+from charms.reactive import set_state, when, when_not
+
+@when('bigtop.available')
+@when_not('zookeeper.installed')
+def install_zookeeper():
+    '''
+    After Bigtop has done the initial setup, trigger a puppet install,
+    via our Zooekeeper library.
+
+    puppet will start the service, as a side effect.
+
+    '''
+    zk = Zookeeper()
+    zk.install()
+    zk.open_ports()
+    set_state('zookeeper.installed')
+    set_state('zookeeper.started')
+    hookenv.status_set('active', 'Ready')
+
+@when('zookeeper.started', 'config.changed.rest')
+def rest_config():
+    """TODO"""
+    pass
+
+@when('zookeeper.started', 'zkpeer.joined')
+def quorum_add(zkpeer):
+    """Add a zookeeper peer.
+
+    Add the unit that just joined, restart Zookeeper, and remove the
+    '.joined' state so we don't fall in here again (until another peer joins).
+    """
+    nodes = zkpeer.get_nodes()  # single node since we dismiss .joined below
+    zk = Zookeeper()
+    zk.increase_quorum(nodes)
+    #restart_zookeeper_if_config_changed()  # Not necessary due to puppet
+    zkpeer.dismiss_joined()
```

Comment:
Indeed, the process is up but the Zookeeper instances do not form a quorum.

If you try to connect to zk with zkCli.sh you fail with:
2016-06-02 11:25:32,733 [myid:] - INFO  [main-SendThread(localhost:2181):ClientCnxn$SendThread@852] - Socket connection established to localhost/127.0.0.1:2181, initiating session
2016-06-02 11:25:32,738 [myid:] - INFO  [main-SendThread(localhost:2181):ClientCnxn$SendThread@1098] - Unable to read additional data from server sessionid 0x0, likely server has closed socket, closing socket connection and attempting reconnect
[zk: localhost:2181(CONNECTING) 0] 2016-06-02 11:25:34,803 [myid:] - INFO  [main-SendThread(localhost:2181):ClientCnxn$SendThread@975] - Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error)

In /var/log/zookeeper/zookeeper.log we get this exception:
2016-06-02 11:31:30,338 [myid:0] - WARN  [RecvWorker:0:QuorumCnxManager$RecvWorker@780] - Connection broken for id 0, my id = 0, error = 
java.io.EOFException
        at java.io.DataInputStream.readInt(DataInputStream.java:392)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager$RecvWorker.run(QuorumCnxManager.java:765)
2016-06-02 11:31:30,338 [myid:0] - WARN  [RecvWorker:0:QuorumCnxManager$RecvWorker@783] - Interrupting SendWorker
2016-06-02 11:31:30,339 [myid:0] - WARN  [SendWorker:0:QuorumCnxManager$SendWorker@697] - Interrupted while waiting for message on queue
java.lang.InterruptedException
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2088)
        at java.util.concurrent.ArrayBlockingQueue.poll(ArrayBlockingQueue.java:418)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager.pollSendQueue(QuorumCnxManager.java:849)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager.access$500(QuorumCnxManager.java:64)
        at org.apache.zookeeper.server.quorum.QuorumCnxManager$SendWorker.run(QuorumCnxManager.java:685)

There are a couple of things you should investigate including if we are affected by this: https://issues.apache.org/jira/browse/ZOOKEEPER-2106

For sure this is a blocker :(